src/cmd/variable.js

```diff
@@ -87,86 +81,81 @@ class VariableCommand {
-			return this.listVariables();
-		} else if (deviceId && !variableName) {
-			//try to figure out if they left off a variable name, or if they want to pull a var from all devices.
-			return this.disambiguateGetValue({ deviceId }).then(({ deviceIds, variableName }) => {
-				return this._getValue(deviceIds, variableName);
-			});
-		} else if (deviceId === 'all' && variableName) {
-			return this.disambiguateGetValue({ variableName }).then(({ deviceIds, variableName }) => {
-				return this._getValue(deviceIds, variableName);
-			});
-		}
+	getValue(deviceId, variableName, { time }) {
+		return Promise.resolve().then(() => {
+			if (!deviceId && !variableName) {
+				//they just didn't provide any args...
+				return this.listVariables();
+			} else if (deviceId && !variableName) {
+				//try to figure out if they left off a variable name, or if they want to pull a var from all devices.
+				return this.disambiguateGetValue({ deviceId }).then(({ deviceIds, variableName }) => {
+					return this._getValue(deviceIds, variableName, { time });
+				});
+			} else if (deviceId === 'all' && variableName) {
+				return this.disambiguateGetValue({ variableName }).then(({ deviceIds, variableName }) => {
+					return this._getValue(deviceIds, variableName, { time });
+				});
+			}
 
-		return this._getValue(deviceId, variableName);
+			return this._getValue(deviceId, variableName, { time });
+		}).catch(err => {
+			const api = new ApiClient();
+			throw new VError(api.normalizedApiError(err), 'Error while reading value');
+		});
 	}
 
 	getAllVariables() {
 		if (this._cachedVariableList) {
-			return when.resolve(this._cachedVariableList);
+			return Promise.resolve(this._cachedVariableList);
 		}
 
 		console.error('polling server to see what devices are online, and what variables are available');
 
 		const api = new ApiClient();
-		if (!api.ready()) {
-			return when.reject('Not logged in');
-		}
-
-		return pipeline([
-			() => {
-				return api.listDevices();
-			},
-			(devices) => {
-				if (!devices || (devices.length === 0)) {
-					console.log('No devices found.');
-					this._cachedVariableList = null;
-				} else {
-					const promises = [];
-					for (let i = 0; i < devices.length; i++) {
-						const deviceid = devices[i].id;
-						if (devices[i].connected) {
-							promises.push(api.getAttributes(deviceid));
-						} else {
-							promises.push(when.resolve(devices[i]));
-						}
+		api.ensureToken();
+
+		return Promise.resolve().then(() => {
+			return api.listDevices();
+		}).then(devices => {
+			if (!devices || (devices.length === 0)) {
+				console.log('No devices found.');
+				this._cachedVariableList = null;
+			} else {
+				const promises = [];
+				for (let i = 0; i < devices.length; i++) {
+					const deviceid = devices[i].id;
+					if (devices[i].connected) {
+						promises.push(api.getAttributes(deviceid));
+					} else {
+						promises.push(Promise.resolve(devices[i]));
 					}
+				}
```

Comment:
Lines 138 -> 146

```
				const promises = [];
				for (let i = 0; i < devices.length; i++) {
					const deviceid = devices[i].id;
					if (devices[i].connected) {
						promises.push(api.getAttributes(deviceid));
					} else {
						promises.push(Promise.resolve(devices[i]));
					}
				}
```

const promises = devices
.map(device => {
  return device.connected ? api.getAttributes(device.id) : Promise.resolve(device);
});

Reply:
I don't think you even need the `Promise.resolve`, just return `device` in the second case.